sycl/plugins/unified_runtime/ur/adapters/cuda/usm.cpp

```diff
@@ -0,0 +1,256 @@
+//===--------- usm.cpp - CUDA Adapter ------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===-----------------------------------------------------------------===//
+
+#include <cassert>
+
+#include "common.hpp"
+#include "context.hpp"
+#include "device.hpp"
+#include "event.hpp"
+#include "platform.hpp"
+#include "queue.hpp"
+
+#include <cuda.h>
+
+/// USM: Implements USM Host allocations using CUDA Pinned Memory
+///
+UR_APIEXPORT ur_result_t UR_APICALL
+urUSMHostAlloc(ur_context_handle_t hContext, const ur_usm_desc_t *pUSMDesc,
+               ur_usm_pool_handle_t pool, size_t size, void **ppMem) {
+  UR_ASSERT(ppMem, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(hContext, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+
+  size_t device_max_mem_alloc_size = 0;
+  UR_ASSERT(urDeviceGetInfo(hContext->get_device(),
+                            UR_DEVICE_INFO_MAX_MEM_ALLOC_SIZE, sizeof(size_t),
+                            static_cast<void *>(&device_max_mem_alloc_size),
+                            nullptr) == UR_RESULT_SUCCESS,
+            UR_RESULT_ERROR_INVALID_DEVICE);
+  UR_ASSERT(size > 0 && size <= device_max_mem_alloc_size,
+            UR_RESULT_ERROR_INVALID_USM_SIZE);
+
+  ur_result_t result = UR_RESULT_SUCCESS;
+  try {
+    ScopedContext active(hContext);
+    result = UR_CHECK_ERROR(cuMemAllocHost(ppMem, size));
+  } catch (ur_result_t error) {
+    result = error;
+  }
+
+  UR_ASSERT(!pUSMDesc || (pUSMDesc->align == 0 ||
```

Comment:
Why is the requested aligment check after the allocation?

Also, if it fails there's a memory leak.

Also happens in device alloc and share alloc

Reply:
Given we were only tasked with moving code from one place to another without causing regressions, could this be pulled out into an issue to be fixed later?

Reply:
Makes sense. I have a patch locally I'll mentally queue